cmd/runtimetest/main.go

```diff
@@ -410,201 +578,379 @@ func validateDefaultFS(spec *rspec.Spec, t *tap.T) error {
+			spec.Version,
+			fmt.Sprintf("%s (default device)", device.Path))
 		if err != nil {
-			if os.IsNotExist(err) {
-				return specerror.NewError(specerror.DefaultDevices,
-					fmt.Errorf("device node %v not found", device),
-					rspec.Version)
-			}
 			return err
 		}
-		if fi.Mode()&os.ModeDevice != os.ModeDevice {
-			return specerror.NewError(specerror.DefaultDevices,
-				fmt.Errorf("file %v is not a device as expected", device),
-				rspec.Version)
-		}
 	}
 
 	return nil
 }
 
-func validateMaskedPaths(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateMaskedPaths(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.MaskedPaths == nil {
+		c.harness.Skip(1, "linux.maskedPaths not set")
 		return nil
 	}
+
 	for _, maskedPath := range spec.Linux.MaskedPaths {
-		f, err := os.Open(maskedPath)
-		if err != nil {
+		readable, err := testReadAccess(maskedPath)
+		if err != nil && !os.IsNotExist(err) {
 			return err
 		}
-		defer f.Close()
-		b := make([]byte, 1)
-		_, err = f.Read(b)
-		if err != io.EOF {
-			return fmt.Errorf("%v should not be readable", maskedPath)
-		}
+		c.harness.Ok(!readable, fmt.Sprintf("cannot read masked path %q", maskedPath))
 	}
+
 	return nil
 }
 
-func validateSeccomp(spec *rspec.Spec, t *tap.T) error {
+func (c *complianceTester) validateSeccomp(spec *rspec.Spec) error {
 	if spec.Linux == nil || spec.Linux.Seccomp == nil {
+		c.harness.Skip(1, "linux.seccomp not set")
 		return nil
 	}
+
 	for _, sys := range spec.Linux.Seccomp.Syscalls {
 		if sys.Action == "SCMP_ACT_ERRNO" {
 			for _, name := range sys.Names {
 				if name == "getcwd" {
 					_, err := os.Getwd()
 					if err == nil {
-						t.Diagnostic("getcwd did not return an error")
+						c.harness.Skip(1, "getcwd did not return an error")
 					}
 				} else {
-					t.Skip(1, fmt.Sprintf("%s syscall returns errno", name))
+					c.harness.Skip(1, fmt.Sprintf("%s syscall returns errno", name))
 				}
 			}
 		} else {
-			t.Skip(1, fmt.Sprintf("syscall action %s", sys.Action))
+			c.harness.Skip(1, fmt.Sprintf("syscall action %s", sys.Action))
 		}
 	}
+
 	return nil
 }
 
-func validateROPaths(spec *rspec.Spec, t *tap.T) error {
-	if spec.Linux == nil {
+func (c *complianceTester) validateROPaths(spec *rspec.Spec) error {
+	if spec.Linux == nil || spec.Linux.ReadonlyPaths == nil {
+		c.harness.Skip(1, "linux.readonlyPaths not set")
 		return nil
 	}
-	for _, v := range spec.Linux.ReadonlyPaths {
-		err := testWriteAccess(v)
-		if err == nil {
-			return fmt.Errorf("%v should be readonly", v)
+
+	for i, path := range spec.Linux.ReadonlyPaths {
+		readable, err := testReadAccess(path)
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		if !readable {
+			c.harness.Skip(1, fmt.Sprintf("%q (linux.readonlyPaths[%d]) is not readable", path, i))
+		}
+
+		writable, err := testWriteAccess(path)
+		if err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		if !writable {
```

Comment:
why the `if`? The condition is already in the `c.harness.Ok()` below.

Reply:
> why the `if`? The condition is already in the `c.harness.Ok()` below.

A sloppy copy/paste from the non-fatal readable test a few lines up.  Fixed by dropping this `if` in 5e395e8 → 1e67d33.